website/client/src/components/challenges/challengeDetail.vue

```diff
@@ -433,6 +433,7 @@ export default {
         this.$router.push('/challenges/findChallenges');
         return;
       }
+      this.$store.state.title = `${this.challenge.name} | Challenges | Habitica`;
```

Comment:
Looking good @jbusa22 !  However in order to support the right titles in all languages, these titles should probably go through the `$t` translation function.

If you want to find out which keys to use for the menu text, you can find them all in https://github.com/HabitRPG/habitica/blob/develop/website/client/src/components/header/menu.vue , or you can search for the exact match in the english translation files.

As an example, this line would become: 
```suggestion
      this.$store.state.title = `${this.challenge.name} | ${$t('challenges')} | ${$('habitica')}`;
```

Note that here the keys are basically the same term as the english word, and most seem to follow this pattern, but it's best to check either in the menu or in the files.

Reply:
Yeah, I'm was thinking of implementing a helper function but it seemed like unnecessary abstraction given the string is only on 1 line per component. In order to separate the translation out, we would need to import the i18n file from common/script/i18n into the util file where the createTitle would be. I think it might just be cleaner to keep it in the string but not sure. Let me know what you think.

Reply:
Ah yeah, you're right, the injection of the function is trivial, wasn't really thinking straight (Was confused by the fact that the function lives on `this` in the vue components).

The introduction of the abstraction is not about reducing lines, it's much more about codifying the fact that all these titles follow the same logic.  That way you decouple the structure of the title from the content of the title.  This way, the implicit knowledge is made explicit and changing the structure of the title in the future becomes easier. 

Technically I think the fact that this solution has duplication (especially in the grouping level) with the menu definition means that this is not really the right place to define it.  It suggests to me that there should be a data structure that holds the grouping information of the links together with the components they link to and the knowledge of how to construct titles based on that grouping and page's content.  The menu can then be rewritten to just generate all the menu items based off of this data structure and inject the titles into the page components also from that data structure (Or from the menu, kinda depends on how you structure it).  That would be a fairly hefty refactoring though, so perhaps as a second step (And definitely only if someone from the core team thinks it's a good idea).

Reply:
I think the option of creating a store action that takes parameters similar to those suggested by @benkelaar and changes the state is the best way to go, as I said previously in the issue. I completely agree with the reasoning outline here in the last comment by @benkelaar

Reply:
Regarding the need to inject the translation function you can use `this.$t` anywhere in components so nothing needs to be injected and outside components (for example in a store action) you can just `import i18n from '@/../../common/script/i18n';` and do `i18n.t()`

Reply:
Ok here is what I have so far for the store action. 
`export function setTitle (store, params) {
  if (params.second) {
    store.state.title = `${i18n.t(params.first)} | ${i18n.t(params.second)} | ${i18n.t('habitica')}`;
  } else {
    store.state.title = `${i18n.t(params.first)} | ${i18n.t('habitica')}`;
  }
}`
Questions I have are:
1. What should the params be named?
2. When i18n doesn't find the string it returns 'String X Not Found'. Obviously, we don't want this showing up in the title so I'm not sure how to get around this. One option is to check all the params first and match the return string for something like 'not found'.  Another option is to put this on the caller of the action. ie setTitle({first: {name: 'Helllo's party', translate: false}})

Reply:
One more question is what file should we put this function in? Right now it lives in the common.js in the store folder.

Reply:
The file should be under store/actions/common.js, I think to avoid issues with the translations the action should be directly passed the translated version (and I don't think Habitica needs to be translated with `i18n.t('habitica')`, let's just use Habitica in english). If a string is missing (String X Not Found) it should be added to `common/locales/en`

Reply:
The reason I thought you might want to translate Habitica I thought might be for languages that use different alphabets, or do you want to stylize the brandname also in roman characters in those languages?

Come to think about it, what about right-to-left languages? Shouldn't those display the title right to left as well? (Perhaps this might be scope creep, probably better to leave to a follow-up story (if at all necessary))

Reply:
I think we want to keep Habitica in latin characters regardless of the user language and regarding RTL languages there's definitely work to be done but we want to tackle that all together and right now it's not in the plans since it would be a big project